### ISSUE TITLE
[reactive-element] queryAssignedElements is initially empty

### ISSUE THREAD

**Issue by msteller-connyun** [Polymer/lit-html#3706]
### Which package(s) are affected?

Lit Core (lit / lit-html / lit-element / reactive-element)

### Description

`queryAssignedElements` is initially empty when a slot is used.

```
  @queryAssignedElements({ slot: 'foo' })
  fooEl!: Array<HTMLElement>;  
  ```

Add an intital value for that slot: 

```
<simple-greeting name="World"><div slot="foo">foo</div></simple-greeting>
```

`this.fooEl.length` is `0`.

### Reproduction

[Example in Playground](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHksIHF1ZXJ5QXNzaWduZWRFbGVtZW50c30gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuXG5AY3VzdG9tRWxlbWVudCgnc2ltcGxlLWdyZWV0aW5nJylcbmV4cG9ydCBjbGFzcyBTaW1wbGVHcmVldGluZyBleHRlbmRzIExpdEVsZW1lbnQge1xuICBzdGF0aWMgc3R5bGVzID0gY3NzYHAgeyBjb2xvcjogYmx1ZSB9YDtcblxuICBAcHJvcGVydHkoKVxuICBuYW1lID0gJ1NvbWVib2R5JztcbiAgXG4gIEBxdWVyeUFzc2lnbmVkRWxlbWVudHMoeyBzbG90OiAnZm9vJyB9KVxuICBmb29FbCE6IEFycmF5PEhUTUxFbGVtZW50PjsgIFxuICBcblxuICByZW5kZXIoKSB7XG4gICAgcmV0dXJuIGh0bWxgPHA-SGVsbG8sICR7dGhpcy5uYW1lfSE8L3A-PHNsb3QgbmFtZT1cImZvb1wiPjwvc2xvdD4ke3RoaXMuZm9vRWwubGVuZ3RofWA7XG4gIH1cbn1cbiJ9LHsibmFtZSI6ImluZGV4Lmh0bWwiLCJjb250ZW50IjoiPCFET0NUWVBFIGh0bWw-XG48aGVhZD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9zaW1wbGUtZ3JlZXRpbmcuanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8c2ltcGxlLWdyZWV0aW5nIG5hbWU9XCJXb3JsZFwiPjxkaXYgc2xvdD1cImZvb1wiPmZvbzwvZGl2Pjwvc2ltcGxlLWdyZWV0aW5nPlxuPC9ib2R5PlxuIn0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMS4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4yLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ)

### Workaround

Triggering manually a `document.querySelector('simple-greeting').requestUpdate()` gives the expected result.

### Is this a regression?

No or unsure. This never worked, or I haven't tried before.

### Affected versions

2.6.1

### Browser/OS/Node environment

-

**Comment by justinfagnani** [comment]
Assigned elements are only available once the `<slot>` has rendered, which is after the first update.

It's usually best to look for assigned elements on `slotchange` events so that you pick up any changed children - this naturally gives the right timing for first updates.

**Comment by gipoezcan** [comment]
@justinfagnani then what's the use-case for queryAssignedElements? Can it be made reactive to provide more value?

**Comment by msteller-connyun** [comment]
Yes, in the end I used addtionally to `queryAssignedElements` a `slotchange` to make it work.

```
<slot name="label" @slotchange=${() => this.requestUpdate()}>

```

However, personally for me, `queryAssignedElements` in the current implementation makes less sense. Especially that the examples on 
https://lit.dev/docs/api/decorators/#queryAssignedElements
is with slot. As I user I would expect that would "just work".
In any case, I would recommend at least add the current limitation to the doc.

**Comment by justinfagnani** [comment]
@gipoezcan the decorator is useful when accessed after the slot has rendered, which can be at many different points: update(), firstUpdate(), an event handler, especially for `slotchange` like @msteller-connyun showed.

I think we could have a doc issue opened to clarify that.

**Comment by YoungElPaso** [comment]
@msteller-connyun good solution, bit by this today as well. Docs could be clearer for sure, but not a big deal once you understand the intent.

**Comment by atruskie** [comment]
Just wanted to note that we were surprised by this behaviour too. Even if you want to maintain the lazy-query style API, it might be nice to pair QueryAssignedElements with some kind of "trigger update on slot change" option or other API. 

But if not, just a doc change for noting the utility of the slotchange event would help.
